### PR TITLE
Refactor pipeline validation to extract reusable code

### PR DESCRIPTION
**Pull request by greeshmaswaminathan**
- Refactor pipeline validation to extract reusable code
- Follow up PR will use the common code for remote execution

### CODE REVIEW

File: cdap-app-templates/cdap-etl/cdap-data-pipeline-base/src/main/java/io/cdap/cdap/datapipeline/service/ValidationHandler.java [hunk 1]
```diff
@@ -91,6 +81,11 @@ public void healthCheck(HttpServiceRequest request, HttpServiceResponder respond
   @Path("v1/contexts/{context}/validations/stage")
   public void validateStage(HttpServiceRequest request, HttpServiceResponder responder,
                             @PathParam("context") String namespace) throws IOException, AccessException {
+    validateLocally(request, responder, namespace);
+  }
+
+  private void validateLocally(HttpServiceRequest request, HttpServiceResponder responder,
```
Comment (tivv) [inline cdap-app-templates/cdap-etl/cdap-data-pipeline-base/src/main/java/io/cdap/cdap/datapipeline/service/ValidationHandler.java:88]: Why do you need to split it? I don't see this reused

File: cdap-app-templates/cdap-etl/cdap-data-pipeline-base/src/main/java/io/cdap/cdap/datapipeline/service/ValidationUtils.java [hunk 2]
```diff
@@ -0,0 +1,88 @@
...
+ * Utility functions for common pipeline validation code
+ */
+public class ValidationUtils {
+
+  private static final Gson GSON = new GsonBuilder()
+    .registerTypeAdapter(Schema.class, new SchemaTypeAdapter())
+    .serializeNulls()
+    .create();
+
+  public static String validate(StageValidationRequest validationRequest,
```
Comment (tivv) [inline cdap-app-templates/cdap-etl/cdap-data-pipeline-base/src/main/java/io/cdap/cdap/datapipeline/service/ValidationUtils.java]: Please add javadoc and a test

File: cdap-app-templates/cdap-etl/cdap-data-pipeline-base/src/main/java/io/cdap/cdap/datapipeline/service/ValidationHandler.java [hunk 3]
```diff
@@ -141,32 +119,31 @@ public void validateStage(HttpServiceRequest request, HttpServiceResponder respo
...
-    }
+    MacroEvaluator macroEvaluator = new DefaultMacroEvaluator(new BasicArguments(programArgs), evaluators);
+    MacroParserOptions macroParserOptions = MacroParserOptions.builder()
+      .skipInvalidMacros()
+      .setEscaping(false)
+      .setFunctionWhitelist(evaluators.keySet())
+      .build();
+    return new Function<Map<String, String>, Map<String, String>>() {
+      @Override
+      public Map<String, String> apply(Map<String, String> macroProperties) {
```
Comment (tivv) [inline cdap-app-templates/cdap-etl/cdap-data-pipeline-base/src/main/java/io/cdap/cdap/datapipeline/service/ValidationHandler.java]: Could you make it a lambda?

**Review by MEseifan — APPROVED** [review]

File: cdap-app-templates/cdap-etl/cdap-data-pipeline-base/src/main/java/io/cdap/cdap/datapipeline/service/ValidationHandler.java [hunk 4]
```diff
@@ -141,32 +119,31 @@ public void validateStage(HttpServiceRequest request, HttpServiceResponder respo
...
     }
 
-    // evaluate secure macros
+    Function<Map<String, String>, Map<String, String>> macroFn = getMacroEvaluationFunction(namespace, arguments);
+    String validationResponse = ValidationUtils
+      .validate(validationRequest, getContext().createPluginConfigurer(namespace), macroFn);
+    responder.sendString(validationResponse);
+  }
+
+  private Function<Map<String, String>, Map<String, String>> getMacroEvaluationFunction(
```
Comment (tivv) [inline cdap-app-templates/cdap-etl/cdap-data-pipeline-base/src/main/java/io/cdap/cdap/datapipeline/service/ValidationHandler.java]: Please consider extracting a lambda one level up. It's usually easier to follow without functions that return functions.
E.g. make `evaluateMacros(namespace, programArgs, macroProperties)` and then  
`macroFn = macroProperties -> evaluateMacros(namespace, arguments, macroProperties)`

File: cdap-app-templates/cdap-etl/cdap-data-pipeline-base/src/main/java/io/cdap/cdap/datapipeline/service/ValidationHandler.java (hunk 1 above)
Comment (greeshmaswaminathan) [reply cdap-app-templates/cdap-etl/cdap-data-pipeline-base/src/main/java/io/cdap/cdap/datapipeline/service/ValidationHandler.java:88]: There will be a config check introduced to put remote execution behind a flag. There will be validateLocally and validateRemotely methods which I thought would give more readability.

File: cdap-app-templates/cdap-etl/cdap-data-pipeline-base/src/main/java/io/cdap/cdap/datapipeline/service/ValidationUtils.java [hunk 5]
```diff
@@ -0,0 +1,88 @@
...
+
+import java.util.Collections;
+import java.util.Map;
+import java.util.function.Function;
+import java.util.function.Supplier;
+
+/**
+ * Utility functions for common pipeline validation code
+ */
+public class ValidationUtils {
```
Comment (chtyim) [inline cdap-app-templates/cdap-etl/cdap-data-pipeline-base/src/main/java/io/cdap/cdap/datapipeline/service/ValidationUtils.java]: final and private constructor for util class.

File: cdap-app-templates/cdap-etl/cdap-data-pipeline-base/src/main/java/io/cdap/cdap/datapipeline/service/ValidationUtils.java [hunk 6]
```diff
@@ -0,0 +1,88 @@
...
+
+  private static final Gson GSON = new GsonBuilder()
+    .registerTypeAdapter(Schema.class, new SchemaTypeAdapter())
+    .serializeNulls()
+    .create();
+
+  public static String validate(StageValidationRequest validationRequest,
+                                PluginConfigurer pluginConfigurer,
+                                Function<Map<String, String>, Map<String, String>> macroFn) {
```
Comment (chtyim) [inline cdap-app-templates/cdap-etl/cdap-data-pipeline-base/src/main/java/io/cdap/cdap/datapipeline/service/ValidationUtils.java]: What does the Function do? If possible, consider use an interface to define a better contract.

File: cdap-app-templates/cdap-etl/cdap-data-pipeline-base/src/main/java/io/cdap/cdap/datapipeline/service/ValidationUtils.java [hunk 7]
```diff
@@ -0,0 +1,88 @@
...
+    // evaluate macros
+    Map<String, String> evaluatedProperties = macroFn.apply(stageConfig.getPlugin().getProperties());
+
+    ETLPlugin originalConfig = stageConfig.getPlugin();
+    ETLPlugin evaluatedConfig = new ETLPlugin(originalConfig.getName(), originalConfig.getType(),
+                                              evaluatedProperties, originalConfig.getArtifactConfig());
+    try {
+      StageSpec spec = pipelineSpecGenerator.configureStage(stageConfig.getName(), evaluatedConfig,
+                                                            pipelineConfigurer).build();
+      return GSON.toJson(new StageValidationResponse(spec));
```
Comment (chtyim) [inline cdap-app-templates/cdap-etl/cdap-data-pipeline-base/src/main/java/io/cdap/cdap/datapipeline/service/ValidationUtils.java]: Generally not a good ideal using a Json string as the method contract. Json serialization should only happen right before writing out the data structure. Why need to do the serialization here?

File: cdap-app-templates/cdap-etl/cdap-data-pipeline-base/src/main/java/io/cdap/cdap/datapipeline/service/ValidationUtils.java (hunk 2 above)
Comment (greeshmaswaminathan) [reply cdap-app-templates/cdap-etl/cdap-data-pipeline-base/src/main/java/io/cdap/cdap/datapipeline/service/ValidationUtils.java]: Done

File: cdap-app-templates/cdap-etl/cdap-data-pipeline-base/src/main/java/io/cdap/cdap/datapipeline/service/ValidationHandler.java (hunk 3 above)
Comment (greeshmaswaminathan) [reply cdap-app-templates/cdap-etl/cdap-data-pipeline-base/src/main/java/io/cdap/cdap/datapipeline/service/ValidationHandler.java]: Done.

File: cdap-app-templates/cdap-etl/cdap-data-pipeline-base/src/main/java/io/cdap/cdap/datapipeline/service/ValidationHandler.java (hunk 4 above)
Comment (greeshmaswaminathan) [reply cdap-app-templates/cdap-etl/cdap-data-pipeline-base/src/main/java/io/cdap/cdap/datapipeline/service/ValidationHandler.java]: Done.

File: cdap-app-templates/cdap-etl/cdap-data-pipeline-base/src/main/java/io/cdap/cdap/datapipeline/service/ValidationUtils.java (hunk 5 above)
Comment (greeshmaswaminathan) [reply cdap-app-templates/cdap-etl/cdap-data-pipeline-base/src/main/java/io/cdap/cdap/datapipeline/service/ValidationUtils.java]: Done

File: cdap-app-templates/cdap-etl/cdap-data-pipeline-base/src/main/java/io/cdap/cdap/datapipeline/service/ValidationUtils.java (hunk 7 above)
Comment (greeshmaswaminathan) [reply cdap-app-templates/cdap-etl/cdap-data-pipeline-base/src/main/java/io/cdap/cdap/datapipeline/service/ValidationUtils.java]: Changed method to return StageValidationResponse

File: cdap-app-templates/cdap-etl/cdap-data-pipeline-base/src/main/java/io/cdap/cdap/datapipeline/service/ValidationUtils.java (hunk 6 above)
Comment (greeshmaswaminathan) [reply cdap-app-templates/cdap-etl/cdap-data-pipeline-base/src/main/java/io/cdap/cdap/datapipeline/service/ValidationUtils.java]: This code can be reused for the remote execution.

**Review by tivv — APPROVED** [review]
nit: make utility class final

**Pull request merged**